File: lib/network/packets/server/s_party.dart
/// 隊伍中的一名成員。
class PartyMemberData {
  const PartyMemberData({
    required this.objId,
    required this.name,
    required this.level,
    required this.hp,
    required this.hpMax,
    required this.mp,
    required this.mpMax,
  });

  final int objId;
  final String name;
  final int level;
  final int hp;
  final int hpMax;
  final int mp;
  final int mpMax;

  factory PartyMemberData.fromJson(Map<String, dynamic> j) => PartyMemberData(
        objId: (j['objId'] as num?)?.toInt() ?? 0,
        name: j['name'] as String? ?? '',
        level: (j['level'] as num?)?.toInt() ?? 1,
        hp: (j['hp'] as num?)?.toInt() ?? 0,
        hpMax: (j['hpMax'] as num?)?.toInt() ?? 0,
        mp: (j['mp'] as num?)?.toInt() ?? 0,
        mpMax: (j['mpMax'] as num?)?.toInt() ?? 0,
      );
}

/// 隊伍狀態。任何變動伺服器都整包重送，所以前端直接換掉整份，不做增量。
///
/// **成員為空代表已離隊或解散** —— 前端據此清空隊伍欄。
class SParty {
  const SParty({
    required this.partyId,
    required this.leaderObjId,
    required this.members,
  });

  final int partyId;
  final int leaderObjId;
  final List<PartyMemberData> members;

  bool get isEmpty => members.isEmpty;

  factory SParty.fromData(Map<String, dynamic> data) => SParty(
        partyId: (data['partyId'] as num?)?.toInt() ?? 0,
        leaderObjId: (data['leaderObjId'] as num?)?.toInt() ?? 0,
        members: [
          for (final m in (data['members'] as List<dynamic>? ?? const []))
            PartyMemberData.fromJson(m as Map<String, dynamic>),
        ],
      );
}

/// 收到組隊邀請。前端應跳出可接受／拒絕的提示。
class SPartyInvite {
  const SPartyInvite({required this.inviterObjId, required this.inviterName});

  final int inviterObjId;
  final String inviterName;

  factory SPartyInvite.fromData(Map<String, dynamic> data) => SPartyInvite(
        inviterObjId: (data['inviterObjId'] as num?)?.toInt() ?? 0,
        inviterName: data['inviterName'] as String? ?? '',
      );
}

File: src/main/java/com/xin/server/packet/server/S_Party.java
package com.xin.server.packet.server;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.xin.server.model.Party;
import com.xin.server.model.instance.PcInstance;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;

/**
 * 隊伍狀態（對應 opcode {@code S_PARTY}）。
 * <p>
 * 任何變動都<b>整包重送</b>給全隊。隊伍最多 5 人，整包重送遠比設計增量封包單純，
 * 也不會出現「某個成員的畫面沒更新到」。
 * <p>
 * 空隊伍（members 為空）代表<b>已離隊或解散</b>，前端據此清空隊伍欄。
 *
 * <pre>
 * {
 *   "op": "S_PARTY",
 *   "data": {
 *     "partyId": 1,
 *     "leaderObjId": 10002,
 *     "members": [
 *       { "objId": 10002, "name": "月蒼海", "level": 12,
 *         "hp": 160, "hpMax": 200, "mp": 30, "mpMax": 100 }
 *     ]
 *   }
 * }
 * </pre>
 */
public class S_Party extends ServerBasePacket {

    private S_Party(Party party) {
        super(ServerOpcodes.S_PARTY);
        put("partyId", party == null ? 0 : party.getId());
        put("leaderObjId",
                party == null || party.getLeader() == null ? 0
                        : party.getLeader().getId());

        ArrayNode arr = newArray();
        if (party != null) {
            for (PcInstance m : party.getMembers()) {
                ObjectNode n = newObject();
                n.put("objId", m.getId());
                n.put("name",  m.getName());
                n.put("level", m.getRealmLevel());
                n.put("hp",    m.getCurrentHp());
                n.put("hpMax", m.getMaxHp());
                n.put("mp",    m.getCurrentMp());
                n.put("mpMax", m.getMaxMp());
                arr.add(n);
            }
        }
        putArray("members", arr);
    }

    public static S_Party of(Party party) {
        return new S_Party(party);
    }

    /** 已離隊／解散：成員為空，前端據此清空隊伍欄。 */
    public static S_Party empty() {
        return new S_Party(null);
    }
}
